src/wrappers/themis/jsthemis/secure_keygen.cpp

```diff
@@ -173,4 +173,100 @@ bool IsPublicKey(const std::vector<uint8_t>& key)
     return false;
 }
 
+Nan::Persistent<v8::Function> SymmetricKey::constructor;
+
+void SymmetricKey::Init(v8::Local<v8::Object> exports)
+{
+    v8::Local<v8::String> className = Nan::New("SymmetricKey").ToLocalChecked();
+
+    // Prepare constructor template
+    v8::Local<v8::FunctionTemplate> thisTemplate = Nan::New<v8::FunctionTemplate>(SymmetricKey::New);
+    thisTemplate->SetClassName(className);
+
+    // Export constructor
+    v8::Local<v8::Function> function = Nan::GetFunction(thisTemplate).ToLocalChecked();
+    constructor.Reset(function);
+    Nan::Set(exports, className, function);
+}
+
+void SymmetricKey::New(const Nan::FunctionCallbackInfo<v8::Value>& args)
+{
+    // If not invoked as "new themis.SymmetricKey(...)" then reinvoke.
+    if (!args.IsConstructCall()) {
+        // We support at most one argument, pass it and ignore others.
+        v8::Local<v8::Value> argv[1] = {args[0]};
+        v8::Local<v8::Function> cons = Nan::New<v8::Function>(constructor);
+        args.GetReturnValue().Set(Nan::NewInstance(cons, args.Length(), argv).ToLocalChecked());
+        return;
+    }
+
+    // If invoked as "new themis.SymmetricKey()" then generate a new key.
+    if (args.Length() == 0) {
+        std::vector<uint8_t> buffer;
+
+        size_t length = 0;
+        themis_status_t status = themis_gen_sym_key(NULL, &length);
+        if (status != THEMIS_BUFFER_TOO_SMALL) {
+            ThrowError("Themis SymmetricKey", status);
+            args.GetReturnValue().SetUndefined();
+            return;
+        }
+
+        buffer.resize(length);
+        status = themis_gen_sym_key(&buffer.front(), &length);
+        if (status != THEMIS_SUCCESS) {
+            ThrowError("Themis SymmetricKey", status);
+            args.GetReturnValue().SetUndefined();
+            return;
+        }
+
+        args.GetReturnValue().Set(CopyIntoBuffer(buffer));
+        return;
+    }
+
+    // If invoked as "new themis.SymmetricKey(value)" then value must be
+    // a byte buffer that we copy.
+    v8::Local<v8::Value> value = args[0];
+    if (!value->IsUint8Array()) {
+        ThrowParameterError("Themis SymmetricKey",
+                            "key is not a byte buffer (use Buffer or Uint8Array)");
+        args.GetReturnValue().SetUndefined();
+        return;
+    }
+    if (node::Buffer::Length(value) == 0) {
+        ThrowParameterError("Themis SymmetricKey", "key is empty");
+        args.GetReturnValue().SetUndefined();
+        return;
+    }
+
+    args.GetReturnValue().Set(CopyIntoBuffer(value));
```

Comment:
if user pass own buffer, why we don't reuse it and return new allocated buffer?
CopyIntoBuffer use nan::CopyBuffer instead nan::NewBuffer which will reuse passed array

Reply:
oh, I re-read comments and understand that we just copy passed buffer, not generate new key... but for what we do this or why it will useful for user? is I understand correctly, that when user pass some Buffer to constructor `themis.SymmetricKey(someBuffer)` we return node::Bufer object with same values?

Reply:
> but for what we do this or why it will useful for user?

Well, we have to copy the buffer so that changes in the original buffer do not affect the key.

As for why this API exists at all, my original idea was that `new themis.SymmetricKey(...)` would return an instance of _SymmetricKey_ — a type that inherits from Buffer, but is not a Buffer. Then it will be possible to restrict Secure Cell interfaces to work only with SymmetricKey instances which are guaranteed to contain only valid keys. In this case the users will need an ability to convert an arbitrary Buffer into SymmetricKey with JsThemis validating the key.

Currently new themis.SymmetricKey() returns instances of buffer, but even if that’s not fixed by the next release I would like to keep this API so that we teach the users to use it instead of using Buffers directly.

Reply:
as I see, `nan` and nodejs buffer api provide two ways to create objects. New created with copy of values `nan::FromBuffer` and second way to reuse user's object `new nan::NewBuffer`. What about to be appropriate with their api and provide two ways too?

Reply:
`Nan::NewBuffer()` transfers the ownership over a `char*` buffer to Node.js Buffer object. That’s not what we want here since the original byte buffer is still owned by the original Buffer object. It’s a utility method to transfer buffers created by C++ code into JavaScript without making an unnecessary copy.

When the user does `new SymmetricKey(someBuffer)` it does not transfer someBuffer into SymmetricKey. Unfortunately, JavaScript is not Rust and even if we did reuse and existing Buffer as SymmetricKey, this will not prevent the user from using `someBuffer` again after the call. For example, they could modify the buffer and this will modify the key. I don’t think that this spooky action at the distance is an expected behavior so it’s better to copy the buffer with `Nan::CopyBuffer()` and avoid such issues.

Reply:
It's okay if they know, that they transfer ownership. in such case they use `new SymmetricKey(someBuffer)` just to cast their buffer to correct type to use with themis in the future. If our function doesn't modify this buffer, it's okay if user use previously created object to re-use if he know, that don't need it in a future. For example (js pseudocode):
```
var buffer = Buffer.alloc(32)
var keys = ['32 length key1', '32 length key2']
var data = ['data1', 'data2']
var encrypted = []
for (i:=0; i<data.length; i++){
  var key = new themis.SymmetricKey(keys[i]);
  encryptedData = new themis.SecureCell(key).encrypt(data[i]);
  encrypted.push(encryptedData);
}
```
here is user reuse same buffer and use it just to cast his keys to correct type and use with themis SecureCell api.

But if we can't do like that, so we only can leave as is...

Reply:
The _above_ code will work just fine. That's what a sane developer (i.e., one thinking like me) would probably write if they wanted to encrypt data with different keys.

The current copying implementation makes the following usage impossible:

```javascript
var buffer = Buffer.alloc(32)
var key = new themis.SymmetricKey(buffer) // note that it's out of the loop
var cell = new themis.SecureCell(key)

var data = ['data1', 'data2']
var encrypted = []
for (var i = 0; i < data.length; i++) {
    getSomeKeyDataInto(buffer)
    encrypted.push(cell.encrypt(data[i]))
}
```

While it definitely has a rationale of a sort, I'd consider it premature optimization and, honestly, a sleeping footgun. It looks like the same cell is used for encryption while in fact the key is different.

Now imagine if the Buffer is actually a reference that's kept in a field or some object. And that object caches buffer instances so the same buffer may be reused later for some completely unrelated data. Even if you don't abuse the buffer like the above, it may still be accidentally changed.

That's why I think that copying the buffer is the most reasonable choice here.